src/coreclr/jit/lower.cpp

```diff
@@ -8212,6 +8212,111 @@ void Lowering::ContainCheckBitCast(GenTree* node)
     }
 }
 
+//------------------------------------------------------------------------
+// TryLowerBlockStoreAsGcBulkCopyCall: Lower a block store node as a CORINFO_HELP_ASSIGN_STRUCT call
+//
+// Arguments:
+//    blkNode - The block store node to lower
+//
+bool Lowering::TryLowerBlockStoreAsGcBulkCopyCall(GenTreeBlk* blk)
+{
+    if (comp->opts.OptimizationDisabled())
+    {
+        return false;
+    }
+
+    // Replace STORE_BLK (struct copy) with CORINFO_HELP_ASSIGN_STRUCT which performs
+    // bulk copy for byrefs.
+    const unsigned bulkCopyThreshold = 4;
```

Comment:
From your benchmark I see that with 4 fields, we're strictly better than the existing approach. How does it look like for 3 fields? E.g. if the perf is roughly the same (not better... just not worse), it might still be advantageous to do bulk copy due to code size. Or does it get worse at 3 fields?

(Extrapolating from the benchmark for 4 and 9 fields makes it look like 3 might still be good, but extrapolation is obviously not a measurement.)

Reply:
@MichalStrehovsky the codegen size is worse with 3 fields, mainly due to explicit nullchecks we have to emit. Performance varies between different machines so I decided to go with 4

Reply:
Makes sense, thanks!